qualtran/bloqs/arithmetic/sorting.py

```diff
@@ -11,118 +11,318 @@
+        out = bb.allocate(dtype=QBit())
+        a, b, out = bb.add(GreaterThan(self.bitsize, self.bitsize), a=a, b=b, target=out)
+        out, a, b = bb.add(CSwap(self.bitsize), ctrl=out, x=a, y=b)
+        return {'a': a, 'b': b, 'out': out}
 
 
 @bloq_example
 def _cmp_symb() -> Comparator:
-    n = sympy.Symbol('n')
-    cmp_symb = Comparator(bitsize=n)
+    L = sympy.Symbol('L')
+    cmp_symb = Comparator(L=L)
     return cmp_symb
 
 
 _COMPARATOR_DOC = BloqDocSpec(bloq_cls=Comparator, examples=[_cmp_symb])
 
 
 @frozen
-class BitonicSort(Bloq):
-    r"""Sort k n-bit numbers.
+class ParallelComparators(Bloq):
+    """Given k numbers in [0, L), compare every pair that is `offset` apart.
+
+    Args:
+        L: Upper-bound (excluded) on the input integers.
+        k: size of the input list.
+        offset: compare numbers whose indices are offset apart.
+
+    Registers:
+        xs: input list of numbers.
+        junk (RIGHT): ancilla generated by comparators.
+    """
+
+    L: SymbolicInt
+    k: SymbolicInt
+    offset: SymbolicInt
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature(
+            [
+                Register("xs", BoundedQUInt(bit_length(self.L - 1), self.L), shape=(self.k,)),
+                Register("junk", QBit(), shape=(self.num_comparisons,), side=Side.RIGHT),
+            ]
+        )
+
+    def is_symbolic(self):
+        return is_symbolic(self.L, self.k, self.offset)
+
+    @cached_property
+    def num_comparisons(self) -> SymbolicInt:
+        return self.k // 2
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', xs: 'SoquetT') -> Dict[str, 'SoquetT']:
+        if self.is_symbolic():
+            raise DecomposeTypeError(f"Cannot decompose symbolic {self=}")
 
-    TODO: actually implement the algorithm using comparitor. Hiding ancilla cost
-        for the moment. Issue #219
+        # make mypy happy
+        assert not isinstance(self.k, sympy.Expr)
+        assert not isinstance(self.offset, sympy.Expr)
+        assert isinstance(xs, np.ndarray)
+
+        comp = Comparator(self.L)
+
+        junk = []
+        for start in range(0, self.k, 2 * self.offset):
+            for step in range(self.offset):
+                i, j = start + step, start + step + self.offset
+                if j >= self.k:
+                    break
+                xs[i], xs[j], anc = bb.add(comp, a=xs[i], b=xs[j])
+                junk.append(anc)
+
+        assert len(junk) == self.num_comparisons
+
+        return {'xs': xs, 'junk': np.array(junk)}
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        return {(Comparator(self.L), self.num_comparisons)}
+
+
+@frozen
+class BitonicMerge(Bloq):
+    r"""Merge two sorted sequences of integers.
 
     Args:
-        bitsize: Number of bits used to represent each integer.
+        L: Upper-bound (excluded) on the input integers.
+        k: Number of integers in each half
```

Comment:
Please specify that `k` must be a power of `2`. Also specify the complexity in terms of `T / Toffoli costs`, `ancilla cost` and `circuit depth`. 

Also, would it be a lot more work if we want to support merging two lists of different sizes?

Reply:
It isn't too difficult, but the edge cases need to be dealt with care. I wanted to get the power-of-2 case working first, will implement the general case in a follow up.